modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/SqlQueryProcessor.java

```diff
@@ -317,6 +334,60 @@ public CompletableFuture<ExecutionTarget> forSystemView(ExecutionTargetFactory f
         services.forEach(LifecycleAware::start);
     }
 
+    /** Get primary replicas. */
+    private CompletableFuture<List<PrimaryReplica>> primaryReplicas(int tableId) {
+        int catalogVersion = catalogManager.latestCatalogVersion();
+
+        Catalog catalog = catalogManager.catalog(catalogVersion);
+
+        CatalogTableDescriptor tblDesc = Objects.requireNonNull(catalog.table(tableId), "table");
+
+        CatalogZoneDescriptor zoneDesc = Objects.requireNonNull(catalog.zone(tblDesc.zoneId()), "zone");
+
+        int partitions = zoneDesc.partitions();
+
+        List<CompletableFuture<PrimaryReplica>> result = new ArrayList<>(partitions);
+
+        HybridTimestamp clockNow = clock.now();
+
+        // no need to wait all partitions after pruning was implemented.
+        for (int partitionId = 0; partitionId < partitions; ++partitionId) {
+            ReplicationGroupId partGroupId = new TablePartitionId(tableId, partitionId);
+
+            CompletableFuture<ReplicaMeta> f = placementDriver.awaitPrimaryReplica(
+                    partGroupId,
+                    clockNow,
+                    AWAIT_PRIMARY_REPLICA_TIMEOUT,
+                    SECONDS
+            );
+
+            result.add(f.thenApply(primaryReplica -> {
+                String holder = primaryReplica.getLeaseholder();
+
+                if (holder == null) {
+                    // additional recovery logic is need to be present around here.
+                    throw new IgniteInternalException(Sql.MAPPING_ERR, "Unable to map query, nothing holds the lease");
+                }
+
+                ClusterNode node = clusterSrvc.topologyService().getByConsistentId(holder);
+
+                if (node == null) {
+                    // additional recovery logic is need to be present around here.
+                    throw new IgniteInternalException(Sql.MAPPING_ERR, "Unable to map query, node is lost or offline");
+                }
+
+                return new PrimaryReplica(node, primaryReplica.getStartTime().longValue());
```

Comment:
Seems that there are similar classes with exact the same semantics e.g `PrimaryReplica` and `NodeWithTerm`. Do we really need all of them?

Reply:
I fill the issue:
https://issues.apache.org/jira/browse/IGNITE-20899